### PR TITLE
Businessintelligence

### PR DESCRIPTION
**Pull request by nico24687**
### Description of changes:
- finished all relationship endpoints for Person A (they pass against the spec harness) 
- have refactored the two revenue queries to be more DRY (let me know what you think) 
- we now have a working favourite customer query

### Tags: @anlewi5 @nico24687

### CODE REVIEW

File: app/models/merchant.rb [hunk 1]
```diff
@@ -31,15 +33,50 @@ def self.search_all(params)
...
+  def self.revenue_by_date(params)
+    revenue = revenue_for_merchant(params[:id])
+      .where(invoice_items: { created_at: params["date"] } )
+      .first
+      .revenue
+
+    revenue_to_json(revenue)
+  end
 
+  def self.revenue_for_merchant(merchant_id)
```
Comment (anlewis) [inline app/models/merchant.rb:68]: I really like how you broke this out. Nice job!

File: app/models/merchant.rb [hunk 2]
```diff
@@ -31,15 +33,50 @@ def self.search_all(params)
...
+    .where(id: merchant_id)
+    .merge(Transaction.successful)
+    .group(:id)
+  end
+
+  def self.revenue_to_json(revenue)
     { revenue: revenue / 100.0 }
   end
 
+#  SELECT "merchants".* FROM "merchants" INNER JOIN "invoices" ON "invoices"."merchant_id" = "merchants"."id" INNER JOIN "customers" ON "customers"."id" = "invoices"."customer_id" INNER JOIN "invoices" "invoices_merchants" ON "invoices_merchants"."merchant_id" = "merchants"."id" INNER JOIN "transactions" ON "transactions"."invoice_id" = "invoices_merchants"."id" WHERE "merchants"."id" = 1
```
Comment (anlewis) [inline app/models/merchant.rb:80]: Can we take this out now?

File: app/models/transaction.rb [hunk 3]
```diff
@@ -1,6 +1,11 @@
 class Transaction < ApplicationRecord
   belongs_to :invoice
 
+
```
Comment (anlewis) [inline app/models/transaction.rb:4]: Whitespace

**Review by anlewis — APPROVED** [review]
Looks good! You did a bit of a different thing with your tests than I would have. I probably would have nested my relationship endpoints in with my record endpoints with 'describe'. Not sure about the pros/cons of these, just a thought to consider.

File: app/models/merchant.rb (hunk 2 above)
Comment (nico24687) [reply app/models/merchant.rb:80]: Oh yeah sure. My bad!

**Pull request merged**